### ISSUE TITLE
Cannot backpropagate through DataParallel

### ISSUE THREAD

**Issue by dalegebit** [pytorch/pytorch#2573]
When I run this code:
```
l = nn.Linear(5,5).cuda()
pl = nn.DataParallel(l)
a = Variable(torch.rand(5,5), requires_grad=True).cuda()
pl(a).sum().backward()
print a.grad
```
I sometimes get `None`, while if I let `a` directly pass through `l` everything is normal. 
I test this code with Python 2.7.6 and pytorch version 0.2.0.

**Comment by apaszke** [comment]
It's because `a` is no longer a leaf Variable after calling `.cuda()` on it (its value is a result of a CPU -> GPU transfer). Change 
```python
a = Variable(torch.rand(5,5), requires_grad=True).cuda()
```
to
```python
a = Variable(torch.rand(5,5).cuda(), requires_grad=True)
```

**Comment by dalegebit** [comment]
Thank you for answering such stupid question! Sorry about the hasty sample code, but I did actually encounter similar problems with my model. When I trained it with multi gpus, I got the zero gradient for all parameters. I tried debugging on torch.cuda.comm module, and finally confirmed this is the root of my problem. Here is the segment of code where weird thing happen.

``` 
File torch/cuda/comm.py   
...
[Line 90]
if nccl.is_available(inputs) and inputs[0].get_device() == destination:
    outputs = [result] + [t.new(t.size()) for t in inputs[1:]]
    nccl.reduce(inputs, outputs, root=destination)
    return result
...
``` 
After executing `nccl.reduce(inputs, outputs, root=destination)`, I found all inputs are accumulated into the wrong destination.
```
(Pdb) p outputs
[
 0
 0
 0
⋮ 
 0
 0
 0
[torch.cuda.FloatTensor of size 5167872 (GPU 1)]
, 
1.00000e-12 *
  0.0000
  0.0000
  0.0000
    ⋮   
  0.0000
  0.0000
  0.0000
[torch.cuda.FloatTensor of size 5167872 (GPU 0)]
]
```
Therefore, all the gradient returns become zero. I reckon this may relate to the different interpretations of device id for nccl and torch. Can anyone explain this?

**Comment by dalegebit** [comment]
And would you consider reopening this issue?

**Comment by ngimel** [comment]
Indeed, that's a bug. Can you try changing this line https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/_functions.py#L16 to ```ctx.input_device = 0``` and check if it fixes your issue? If you've installed pytorch with ```setup.py build develop``` changing that file in your source tree should be enough, otherwise, if you don't want to rebuild for a single line change you'd have to find it in the install trees.

**Comment by dalegebit** [comment]
So if I directly set `ctx.input_device` to 0 I can no longer flexibly assign my model to arbitrary devices, right?

**Comment by dalegebit** [comment]
Oh, it throws this error:
`RuntimeError: arguments are located on different GPUs at /pytorch/torch/lib/THC/generated/../generic/THCTensorMathPointwise.cu:558`

**Comment by ngimel** [comment]
Sorry about that. Can you post a minimum reproducer that accumulates results on the wrong GPU?

**Comment by dalegebit** [comment]
Sure.
```
import torch
from torch.autograd import Variable

torch.cuda.set_device(1)
a = Variable(torch.rand(5,6).cuda(), requires_grad=True)
l = torch.nn.Linear(6,7).cuda()
p_l = torch.nn.DataParallel(l, [1,0])
p_l(a).sum().backward()
print l.weight.grad
```
I found that directly setting this line [https://github.com/pytorch/pytorch/blob/master/torch/cuda/comm.py#L92](url) to `nccl.reduce(inputs, outputs, root=0)` temporarily fixes my problem.

**Comment by ngimel** [comment]
Yes, setting ctx.input_device also should have propagated 0 as the root value for nccl call. You should still be able to assign your model to arbitrary device, 0 is not device id, it is the 0-th element of the device_id list that was used to create nccl communicator, which corresponds to device 1 in your case.

**Comment by dalegebit** [comment]
Indeed. Thank you!

**Comment by soumith** [comment]
fixed via #2632
